js-jest/src/gilded_rose.js

```diff
@@ -9,201 +9,142 @@ class Item {
-
-      if (this.items[index].quality > 50){
-        this.items[index].quality = 50
-      }
+}
 
-      return this.items[index]
-    }
+class LegendaryItem extends Item {
+  constructor(name, sellIn, quality){
+    super(name, sellIn, quality)
   }
 
-  _updateLegendaryItem(index){
-    this.items[index].sellIn = 0
+  updateItem(){
+    this.sellIn = 0
 
-    if(this.items[index].quality > 50){
-      this.items[index].quality = 50
+    if(this.quality > 50){
+      this.quality = 50
     }
+  }
+}
 
-    return this.items[index]
+class ConcertTickets extends Item {
+  constructor(name,sellIn, quality){
+    super(name,sellIn, quality)
   }
 
-  _updateConcertTickets(index){
-    const sellBy = this.items[index].sellIn
+  updateItem(){
+    const sellBy = this.sellIn
+
     switch(true){
       case (sellBy <= 0): 
-        this.items[index].quality = 0
+        this.quality = 0
         break
 
       case (sellBy > 0 && sellBy <= 5):
-        this.items[index].quality += 3
-        if (this.items[index].quality > 50){
-          this.items[index].quality = 50
+        this.quality += 3
+        if (this.quality > 50){
+          this.quality = 50
         }
         break
 
       case (sellBy > 5 && sellBy <= 10):
-        this.items[index].quality +=2
-        if (this.items[index].quality > 50){
-          this.items[index].quality = 50
+        this.quality +=2
+        if (this.quality > 50){
+          this.quality = 50
         }
         break
 
       case (sellBy > 10):
-        this.items[index].quality += 1
-        if (this.items[index].quality > 50){
-          this.items[index].quality = 50
+        this.quality += 1
+        if (this.quality > 50){
+          this.quality = 50
         }
         break
     }
   }
+}
 
-  updateQuality() {
-    for (let i = 0; i < this.items.length; i++) {
-      const itemName = this.items[i].name
-
-      switch(true){
-        case (itemName === 'Backstage passes to a TAFKAL80ETC concert'):
-          this._updateConcertTickets(i)
-          break
-
-        case (itemName === "Sulfuras, Hand of Ragnaros"):
-          this._updateLegendaryItem(i)
-          break
+class ConjuredItem extends Item {
+  constructor(name, sellIn, quality){
+    super(name, sellIn, quality)
+  }
 
-        case (itemName === 'Aged Brie'):
-          this._updateAgedBrie(i)
-          break
+  updateItem(){
+    if (this.sellIn >0){
+      this.quality -= 2
 
-        case (itemName === "Conjured Item"):
-          this._updateConjuredItem(i)
-          break
+      if(this.quality < 0){
+        this.quality = 0
+      }
+    } else{
+      this.quality -= 4
 
-        default:
-          this._updateNormalItem(i)
-          break
+      if(this.quality < 0){
+        this.quality = 0
       }
     }
+  }
+}
 
-    return this.items;
+class Shop {
+  constructor(items=[]){
+    this.items = items;
+  }
+
+  updateQuality() {
+    this.items.forEach(item => {
+      item.updateItem()
+      item.sellIn -= 1})
```

Comment:
The LegendaryItem not conforming to this abstraction is kind of annoying right?
To me, it feels weird to have to set its quality to 0 each time `updateItem` is called. So I'd probably move this logic that decrements `sellIn` out of here and into each item type's class.
Also, I did a quick look at the requirements and I think Sulfuras is supposed to have a quality of 80 always.

Reply:
Talked IRL and agreed that `sellIn` is ok to leave as-is at this point. Still want to fix the 80 though.